Когда python код готовится к выполнению, строится синтаксическое дерево. В это время идет проверка структуры кода. Ошибки в структуре кода, например
```python
if True
	print("True")
```
Видны сразу и программа даже не запустится
![[Pasted image 20250602224906.png]]
НО! Если ошибка будет менее очевидна, например, обращение к переменной внутри функции, которой нет, то все будет работать, пока функция не вызовется
```python
def f():  
    print(a)  
    a = 20  
  
  
print("Hello")  # Hello
```
Но вот если мы эту функцию вызовем, то будет уже ошибка
```python
def f():  
    print(a)  
    a = 20  
  
  
print("Hello")  # Hello
print(f()) # UnboundLocalError: local variable 'a' referenced before assignment
```
Все из-за динамической природы Python :)
